services/repository/push.go

```diff
@@ -0,0 +1,365 @@
+	repo_module "code.gitea.io/gitea/modules/repository"
+	"code.gitea.io/gitea/modules/setting"
+	pull_service "code.gitea.io/gitea/services/pull"
+)
+
+// PushUpdateOptions defines the push update options
+type PushUpdateOptions struct {
+	PusherID     int64
+	PusherName   string
+	RepoUserName string
+	RepoName     string
+	RefFullName  string
+	OldCommitID  string
+	NewCommitID  string
+}
+
+// IsNewRef return true if it's a first-time push to a branch, tag or etc.
+func (opts PushUpdateOptions) IsNewRef() bool {
+	return opts.OldCommitID == git.EmptySHA
+}
+
+// IsDelRef return true if it's a deletion to a branch or tag
+func (opts PushUpdateOptions) IsDelRef() bool {
+	return opts.NewCommitID == git.EmptySHA
+}
+
+// IsUpdateRef return true if it's an update operation
+func (opts PushUpdateOptions) IsUpdateRef() bool {
+	return !opts.IsNewRef() && !opts.IsDelRef()
+}
+
+// IsTag return true if it's an operation to a tag
+func (opts PushUpdateOptions) IsTag() bool {
+	return strings.HasPrefix(opts.RefFullName, git.TagPrefix)
+}
+
+// IsNewTag return true if it's a creation to a tag
+func (opts PushUpdateOptions) IsNewTag() bool {
+	return opts.IsTag() && opts.IsNewRef()
+}
+
+// IsDelTag return true if it's a deletion to a tag
+func (opts PushUpdateOptions) IsDelTag() bool {
+	return opts.IsTag() && opts.IsDelRef()
+}
+
+// IsBranch return true if it's a push to branch
+func (opts PushUpdateOptions) IsBranch() bool {
+	return strings.HasPrefix(opts.RefFullName, git.BranchPrefix)
+}
+
+// IsNewBranch return true if it's the first-time push to a branch
+func (opts PushUpdateOptions) IsNewBranch() bool {
+	return opts.IsBranch() && opts.IsNewRef()
+}
+
+// IsUpdateBranch return true if it's not the first push to a branch
+func (opts PushUpdateOptions) IsUpdateBranch() bool {
+	return opts.IsBranch() && opts.IsUpdateRef()
+}
+
+// IsDelBranch return true if it's a deletion to a branch
+func (opts PushUpdateOptions) IsDelBranch() bool {
+	return opts.IsBranch() && opts.IsDelRef()
+}
+
+// TagName returns simple tag name if it's an operation to a tag
+func (opts PushUpdateOptions) TagName() string {
+	return opts.RefFullName[len(git.TagPrefix):]
+}
+
+// BranchName returns simple branch name if it's an operation to branch
+func (opts PushUpdateOptions) BranchName() string {
+	return opts.RefFullName[len(git.BranchPrefix):]
+}
+
+// RepoFullName returns repo full name
+func (opts PushUpdateOptions) RepoFullName() string {
+	return opts.RepoUserName + "/" + opts.RepoName
+}
+
+// pushQueue represents a queue to handle update pull request tests
+var pushQueue queue.Queue
+
+// handle passed PR IDs and test the PRs
+func handle(data ...queue.Data) {
+	for _, datum := range data {
+		opts := datum.([]*PushUpdateOptions)
+		if err := pushUpdates(opts); err != nil {
+			log.Error("pushUpdate failed: %v", err)
+		}
+	}
+}
+
+func initPushQueue() error {
+	pushQueue = queue.CreateQueue("push_update", handle, []*PushUpdateOptions{}).(queue.Queue)
```

Comment:
There are defaults that maybe we should want to change, here. Like `BATCH_LENGTH = 20`.

Reply:
This could be changed by user.

Reply:
What I mean is that it might be useful to tailor these defaults. We're currently using the same queue defaults for all queues, but maybe git-centered activity queues could benefit from defaults different than e-mail delivery activity queues, for example.

Reply:
If this is wanted specific defaults overrides would be set:

https://github.com/go-gitea/gitea/blob/e4b3f35b8d68d6409a280a8e644759e10b091cb1/modules/setting/queue.go#L87-L163